source/GLaDOS.cpp

```diff
@@ -276,9 +392,112 @@ void GLaDOS::update(float dt) {
 		readyQueue.erase(readyQueue.begin() + i);
 		break;
 	}
+}
 
-	if (fail) {
-		mib->failAllTask();
-		fail = false;
+void GLaDOS::tutorialLevels(float dt) {
+	switch (levelNum) {
+		case tutorial::BREACH_LEVEL:
+			if (ship->timePassed() >= tutorial::B_L_PART1 && things == 2) {
+				CULog("things : %d", things);
+				float actualWidth = ship->getSize() / (float)sections;
+				float width = actualWidth - tutorial::FAKE_DOOR_PADDING * 2;
+				for (int i = 0; i < ship->getDonuts().size(); i++) {
+					float suggestedAngle =
+						ship->getDonuts().at(i)->getAngle() + tutorial::BREACH_DIST;
+					if (suggestedAngle >= ship->getSize()) suggestedAngle -= ship->getSize();
+					float mid = actualWidth * (float)i;
+					float diff =
+						ship->getSize() / 2 - abs(abs(suggestedAngle - mid) - ship->getSize() / 2);
+					if (diff > width / 2) {
+						// clamp this angle within the width of the section
+						suggestedAngle -= width;
+						if (suggestedAngle < 0) suggestedAngle += ship->getSize();
+					}
+					placeObject({BuildingBlockModel::Breach, 0, -1}, suggestedAngle,
+								(i + 1) % ship->getDonuts().size());
+				}
+				things--;
```

Comment:
What is `things`? Could it get a more descriptive name?

Reply:
well it's described in the associated documentation/comment, but I'll change it just for you